pkg/controller/core/admissioncheck_controller.go

```diff
@@ -220,7 +222,8 @@ func (r *AdmissionCheckReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	}
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&kueue.AdmissionCheck{}).
+		WithOptions(controller.Options{NeedLeaderElection: ptr.To(false)}).
 		WatchesRawSource(&source.Channel{Source: r.cqUpdateCh}, &handler).
 		WithEventFilter(r).
-		Complete(r)
+		Complete(WithLeadingManager(mgr, r))
```

Comment:
My understanding is only controllers that contribute to the `queue.Manager` state, which is the source of truth for the visibility API, have to be adapted so their watchers are started unconditionally (w.r.t. leader election).

Reply:
I think this SGTM, but let me clarify what I know and ask some questions to better understand, if the issue is just with the visibility API server, or with the scheduling decisions in the HA mode.

IIUC, the Visibility API serves the information from queue.Manager, but we accept there is some delay in the propagation of the information between replicas. 

What is bad is if the information is not just delayed, but wrong? If so, could it go wrong by running multiple replicas of the core reconcilers?

Further, if the information in the queue.Manager is wrong (for a selected replica), then could this also lead to wrong scheduling decisions? If so, then we have a bigger problem than just "visibility API consistency", but correctness of the scheduling decisions.

Reply:
With the approach taken in this PR, only the leading replica is still responsible for making scheduling decisions, based on the state of its local queue manager. So this does not change the scheduling behaviour, compared to the single replica scenario.

From the implementation details perspective, the reconcilers are "disabled" for the non-leading replicas, and only the watchers are started, so no "write" changes are made to the resources. Only the state of the queue managers is updated, which only impact the results of the visibility API requests made to these replicas when these are non-leading.

Given the asynchronous nature of the watchers, there may be a delay by the time they all eventually reach the stationary state of the system, which can result in discrepancies during the transition. If some events are missed, there should be a full re-sync, as it's done for the informers. We may have to harden that re-sync mechanism, if it turns out to be a real issue.

Reply:
Thank you for the explanations, I think the PR makes sense, I would like to better understand the status on the current main.

Is the issue (on the current main) just with the visibility API serving inconsistent data (which could be anyway due to delays), or the issue is with scheduling decisions? If the issue is with making scheduling decisions then your PR fixes more than just visibility API.

Reply:
cc @alculquicondor

Reply:
> Is the issue (on the current main) just with the visibility API serving inconsistent data (which could be anyway due to delays), or the issue is with scheduling decisions?

The issue with current main is the fact that visibility API endpoints that are served by the non-leading replicas return empty results. That's because the watchers that update the queue manager aren't started for these replicas, which is what that PR solves.

For the correctness of scheduling decisions, this stays untouched, as it remains the responsibility of the leading replica alone.

Reply:
Thanks for explaining, so I guess the only remaining question from me would be where in this change we actually start the watchers in this PR: `That's because the watchers that update the queue manager aren't started for these replicas, which is what that PR solves.`?

Reply:
You're welcome.

The watchers are started for non-leading replicas with the addition of `WithOptions(controller.Options{NeedLeaderElection: ptr.To(false)})`.

This actually instructs controller-runtime to start the entire controllers. When a controller starts, both its reconciler and watchers are started. The other addition, `WithLeadingManager(mgr, r)`, prevents the reconciler from doing any actual work for non-leading replicas, until one acquire the leader election lease possibly.

Reply:
Excellent, got it, the meaning of `WithOptions(controller.Options{NeedLeaderElection: ptr.To(false)}).` was the missing piece for me. Thinking if there is a good place to note it, maybe summarize this in the PR description in a bit more detail.

Reply:
I've updated the PR description as you've suggested. Let me know if there is any other places I could capture this.

Reply:
sgtm, proposed in a separate comment